README.md
# PythonDXCC Files Area 13 Nov 2017

This app is being developed, new feature will be added as time allows!

A lot of logging programs have this capability, some far more, however these programs can be very large and do
things their way.  I wanted a simple application, portable across multiple platforms that can do this simple 
task for me.  If you find it useful by all means use it or modify it for you own use.  
This may morph into a complete simple logger with its operation geared for DX'ing and possible protable operations
with minimal logging, i.e. no callbook lookup, no radio querying.  Just a plain old simple logger, small and cross
platform, ideal for portable or a simpler way of operating.

So far the app integrates LoTW verified ADI file with Country Information from plist file (XML) (http://www.country-files.com)
Allows a Amateur Radio DXCC Awards operator to input a prefix and receive information based on the ARRL's
verification of contacts for DXCC awards.  In an instance you can see if you need the entity you hear for any
DXCC awards.  If you have ARRL DXCC awards for this entity it shows VERIFIED! and the awards you have.  So if you're on 
40m and hear XE2XE (a verified award entity) you can type in XE and it will return Mexico as the entity and show VERIFIED! 
and the awards you currently have verified.  If 40m is not checked... go get em, you need that one! Of course you can also 
have a chat with the other operator just for fun too! Say Mexico has never been verified.. you will see no awards and 
a red NEEDED!  Again go get em, you need this one!  This only applies to ARRL VERIFIED AWARDS not Worked or Confirmed!
Just because you worked the station doesn't mean you will get confirmation and just because you get confirmation (though
QSL card or LoTW) doesn't mean the ARRL will award that contact!  You must apply for the award for the Award file to 
include the contact entity.  To me this is simpler, work them if you NEED them, even if you have worked the entity
before.  You may not get the confirmation or the verification so don't miss the opportunity!

Enhancement in the future..

Currently the app hard codes the ARRL ADIF Verification File, DXCC Entity File and those files have to be 
manually downloaded.  Eventually there will be a section for downloading them automatically if the file is 
not present or after the file reaches a certain age (configurable).  An option to update these files manually
will also be added.


Root Directory Files:
File  PyQtDXCC.py              - (MAIN PROG) This is the main application for this section.  However it will be expanded 
                                 in future to do more things.  Hence the test programs that I have kept in the
                                 archive.                
File DXCCPrefixLookup.ui       - (DATA) QtCreator file for UI (Data Used)
File DXCC_QSLs_20171109_184150 - (DATA) This file is downloaded from the ARRL Log Book of the World (LoTW).
                                 It contains all verified and award connected contacts in ADIF format.
                                 (Data Used)
File  PrefixLookupClass.py     - (PROG TEST) Simple console version of GUI app.  Not completed but works.
File  python_adi.py            - (TEST) Simple parser for testing that I used to generate a text file of aggregate entities.
File  regexp.py                - (TEST) Testing out a regular expression for parsing DXCC Cluster telnet data.
File  TelnetCluster.py         - (TEST) Simple console telnet connection and output.  Lookup to be integrated in future app.
File DXCCPrefixLookup.py       - (TEST) Converted UI file to python code only for reference not used.
File Screenshot.png            - (INFO) A screenshot of the application.
File DXCC200.png               - (INFO) Not used just a screenshot of my DXCC finally reaching 200 entities.

Directory cty_plist Files:
  File  cty.plist         - (DATA) File that contains DXCC entity information.  
                            Thanks to Jim Reisert AD1C for keeping this information current! 
                            Location for download: http://www.country-files.com
                            (Data Used)
  File  history.txt       - What has changed, adds deletes and modifications to DXCC entities.
  File  readme.txt        - Self explanatory!
  
  






cty_plist/history.txt
    30 October 2017 (CTY-2712)
    VER20171030, Version entity is Togo, 5V

    Added/changed Entities/Prefixes/Callsigns:

	* 4U2U is Vienna Intl Ctr, *4U1V
	* GB2ELH is Shetland Islands, *GM/s
	* 3V8CB/J is Tunisia, 3V
	* 9H9PSG/J is Malta, 9H
	* 9M4SJSD is East Malaysia, 9M6
	* GB2GU is Guernsey, GU
	* 4U24OCT, IQ6VP/J, IQ7ML/J and IZ4AIH/J are all Italy, I
	* II0M is Sardinia, IS
	* KD0EE is United States, K in CQ zone 4, ITU zone 8
	* KB3SEV, KC2HL, KC9SBG, KE7PIZ and W7EHP are all Hawaii, KH6
	* KA2ZSD and KC0TK are both Alaska, KL
	* K4D is Puerto Rico, KP4
	* LU1AET/D is Argentina, LU
	* 4U2U is Austria, OE
	* OK1KCR/J is Czech Republic, OK
	* PA1FJ/J, PA4WK/J, PA6HYG/J, PA6JAM/J, PA6SJS/J, PH50GFB/J, PH6BB/J
	  and PI4CQ/J are all Netherlands, PA
	* SV2KBB/J and SZ2TSL/J are both Greece, SV
	* R8MB/1 is European Russia, UA in CQ zone 17, ITU zone 20
	* R3CG/8 and RK6YM/8 are both Asiatic Russia, UA9 in ITU zone 20
	* VK9AR/6 is Australia, VK in CQ zone 29, ITU zone 58
	* YU1BBA/J is Serbia, YU

    Removed Entities/Prefixes/Callsigns:

	* TX5EG in French Polynesia, FO
	* GB0WHR and GB1SPN in Wales, GW
	* AI0M in United States, K
	* KI6EFY and NH2KH in Hawaii, KH6
	* AH0AH, KE7FNC and WA7B in Alaska, KL
	* LU3HPW/H, LU6HBB/Q and LU7AA/U in Argentina, LU
	* R73QRQ, R870B, R8AEU/6 and R9KW/4 in European Russia, UA
	* RA/UR8IF, UA0QQO/9/P and UE70AAA in Asiatic Russia, UA9


regextest.py
import re

telnet_output = 'DX de K7XC:       3525.0  VK9MA        CQ CW CQing up1 Tnx QSO!  VK9M 0748Z NV'

callsign_pattern = "([a-z|0-9|/]+)"
frequency_pattern = "([0-9|.]+)"
pattern = re.compile("^DX de "+callsign_pattern+":\s+"+frequency_pattern+"\s+"+callsign_pattern+"\s+(.*)\s+(\d{4}Z)", re.IGNORECASE)

match = pattern.match(telnet_output)

if match:
	spotter = match.group(1)
	frequency = float(match.group(2))
	spotted = match.group(3)
	comment = match.group(4).strip()
	spot_time = match.group(5)
#	band = frequency_to_band(frequency)

print(spotter, frequency, spotted, comment, spot_time)

'''
I believe that the only way to get the prefix is to go through the DX call and iteriate 
through the call checking the prefix.  an example is VK9MA...

V = Nothing
VK = Australia
VK9 = Norfolk Island
VK9M = Mellish Island 
VK9MA = Nothing

So for this call the prefix is VK9M...  and the entity is Mellish Island
Seems a bit weird to have to do it this way but my poor brain can't see any other
way at the moment. 

Have to watch for stupid calls as well...  PJ4/K9VD would be an example...
No idea about a K9VD/DL1 type of call if that is even possible!

The spot network that I am on gives the prefix for DX calls not US or Canada
for those it give the dang state or provance in the tail end of the comments 
string.  So if I start at the end of the string and get the chars up to the first 
space and compare with the DX call; this should be the prefix.  Will have to test 
all this out or jsut use it and when it fails figure out why.  

Sad nothing is documented in the HAM radio world!
   
''' 


python_adi.py
#!/usr/bin/python3

# Python to parse my LoTW DXCC accepted QSL's (Downloaded from LoTW site) so I can quickly see if I have work 
# an entity or not. So far I do not look at band or mode just if I have worked an entity or not period.
# I need to add if I need a band or mode for the entity as well.  The output needs to show that I 
# have the entity but need different bands or modes.
#
# I am only really intrested in 80m 40m 20m 15m 10m.  160m 30m 17m 12m I will take if I don't have the entity!
# Mode wise for now only Phone and CW.  Digital and Sat I may try later...  I want 5 Band DXCC for Phone and CW! 
#
# NOTE:  This is also only for accepted LoTW credits.  If I have not applied for the new credits
# i.e. new DXCC contacts they will not show in this adi file!  I should keep another file of 
# credits to be applied for as I work them and add them to this program.
#

DXCCTag = "<COUNTRY:"
DXCCDict = {}
DXCCKey = ""
Count = 0;

try:
	f = open("DXCC_QSLs_20171109_184150.adi", encoding = 'utf-8')
	for line in f:
		if line[0:8] == DXCCTag[0:8]:
			index = line.index('>')
			DXCCKey = line[index+1: ].rstrip()
			if DXCCKey not in DXCCDict:
				DXCCDict[DXCCKey] = DXCCKey		# Value needs to contain MODE and BAND confirmed.
				Count += 1;

	print ("No DXCC's imported: ", Count)
	for K in sorted(DXCCDict):					# Need to add the bands and modes worked...
		print (K)
finally:
	f.close()

# APP_LoTW_CREDIT_GRANTED - Award Enumeration
#DXCC-2			 2 Meters
#DXCC-6			 6 Meters
#DXCC-10		10 Meters
#DXCC-12		12 Meters
#DXCC-15		15 Meters
#DXCC-17		17 Meters
#DXCC-20		20 Meters
#DXCC-30		30 Meters
#DXCC-40		40 Meters
#DXCC-80		80 Meters
#DXCC-160		160 Meters
#DXCC-5B		5-Band
#DXCC-5B-6		5-Band (6 Meter Endorsement)
#DXCC-5B-12		5-Band (12 Meter Endorsement)
#DXCC-5B-17		5-Band (17 Meter Endorsement)
#DXCC-5B-30		5-Band (30 Meter Endorsement)
#DXCC-5B-160	5-Band (160 Meter Endorsement)
#DXCC-CHAL		Challenge
#DXCC-CW		CW
#DXCC-CW-HR		CW (Honor Roll)
#DXCC-M			Mixed
#DXCC-M-HR		Mixed (Honor Roll)
#DXCC-PH		Phone
#DXCC-PH-HR		Phone (Honor Roll)
#DXCC-RTTY		Digital
#DXCC-RTTY-HR	Digital (Honor Roll)
#DXCC-SAT		Satellite


PyQtDXCC.py
#!/usr/bin/python3

# pyuic4 -x DXCCPrefixLookup.ui -o DXCCPrefixLookup.py
# Another way is to use the pyuic4 created source file
# this article shows how to connect up events so they 
# are seperate from teh generated file.  That way you don't
# lose work if you redesigne the UI and regen the file.
# https://nikolak.com/pyqt-qt-designer-getting-started/

import sys
import time
import plistlib
from PyQt4 import Qt, QtCore, QtGui, uic

DXCCDict = {}
 
qtCreatorFile = "DXCCPrefixLookup.ui" # Enter ui file here.
 
Ui_MainWindow, QtBaseClass = uic.loadUiType(qtCreatorFile)

class MyApp(QtGui.QMainWindow, Ui_MainWindow):
	def __init__(self):
		QtGui.QMainWindow.__init__(self)
		Ui_MainWindow.__init__(self)
		self.setupUi(self)

		#Connect events here!
		self.lineEditPrefix.returnPressed.connect(self.on_lineEditReturn)

		# all checkboxes are READ ONLY don't allow the user to check them!
		self.checkBox2m.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBox2m.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBox2m.setFocusPolicy(QtCore.Qt.NoFocus)
		
		self.checkBox6m.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBox6m.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBox6m.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBox10m.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBox10m.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBox10m.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBox12m.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBox12m.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBox12m.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBox15m.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBox15m.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBox15m.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBox17m.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBox17m.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBox17m.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBox20m.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBox20m.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBox20m.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBox30m.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBox30m.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBox30m.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBox40m.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBox40m.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBox40m.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBox80m.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBox80m.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBox80m.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBox160m.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBox160m.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBox160m.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBox5Band.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBox5Band.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBox5Band.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBox5B6mE.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBox5B6mE.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBox5B6mE.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBox5B12mE.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBox5B12mE.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBox5B12mE.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBox5B17mE.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBox5B17mE.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBox5B17mE.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBox5B30mE.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBox5B30mE.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBox5B30mE.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBox5B160mE.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBox5B160mE.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBox5B160mE.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBoxMixedHR.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBoxMixedHR.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBoxMixedHR.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBoxMixed.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBoxMixed.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBoxMixed.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBoxPhone.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBoxPhone.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBoxPhone.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBoxSatellite.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBoxSatellite.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBoxSatellite.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBoxDigitalHR.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBoxDigitalHR.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBoxDigitalHR.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBoxCW.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBoxCW.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBoxCW.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBoxCWHR.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBoxCWHR.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBoxCWHR.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBoxPhoneHR.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBoxPhoneHR.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBoxPhoneHR.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBoxChallenge.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBoxChallenge.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBoxChallenge.setFocusPolicy(QtCore.Qt.NoFocus)

		self.checkBoxDigital.setAttribute(QtCore.Qt.WA_TransparentForMouseEvents, True)
		self.checkBoxDigital.setAttribute(QtCore.Qt.WA_NoChildEventsForParent, True)
		self.checkBoxDigital.setFocusPolicy(QtCore.Qt.NoFocus)

		# Start with Awrards section hidden
		self.hideAwardsSection()
		
		# While loading hide all other controls and labels
#		self.labelNumberVerifiedText.hide()
#		self.labelNumberVerifiedCount.hide()
#		self.labelPrefix.hide()
#		self.labelVerified.hide()
#		self.lineEditPrefix.hide()
#		self.labelDXCCPrefixEnter.hide()
		# Use the Country Lable to tell user we are loading country data, it takes a bit.
#		self.labelCountry.setText("Loading DXCC list and ARRL Verified data")


		#some howto do for later!
		self.labelNumberVerifiedCount.setText("200")
		self.labelNumberVerifiedCount.setStyleSheet("color: red")
		self.checkBox160m.setChecked(True)

	# Here is where I do the lookup and display the information!
	def on_lineEditReturn(self):
		inputPrefix = self.lineEditPrefix.text()
		self.lineEditPrefix.setText("")
		if inputPrefix:

			self.labelPrefix.setStyleSheet("color: darkred");
			self.labelPrefix.setText(inputPrefix.upper())
			if inputPrefix.upper() in pl:
				self.labelCountry.setStyleSheet("color: blue; font-style: italic");
				self.labelCountry.setText(pl[inputPrefix.upper()]['Country'])
				# Now check if we are Verified or not
				# If we are, display Verified and set awards checkboxes 
				# and display the awards verified by ARRL
				DXCCId = str (pl[inputPrefix.upper()]['ADIF'])			# Must explicity state this is a string or it will thinkk it is an int! Wow just Wow
				if DXCCDict.get(DXCCId) == None:
					self.labelVerified.setStyleSheet("color: red; font-weight: bold")
					self.labelVerified.setText("NEEDED!")
					self.hideAwardsSection()
				else:
					self.labelVerified.setStyleSheet("color: limegreen; font-weight: bold")
					self.labelVerified.setText("VERIFIED!")
					CurrentRecord = DXCCDict.get(DXCCId)
					self.setAwards(CurrentRecord)
					
  
			else:
				self.hideAwardsSection()
				self.labelCountry.setStyleSheet("color: red; font-weight: bold")
				self.labelCountry.setText("Invalid Prefix")
				self.labelVerified.setText("")	

	def clearAllAwards(self):
		self.checkBox2m.setChecked(False)
		self.checkBox6m.setChecked(False)
		self.checkBox10m.setChecked(False)
		self.checkBox12m.setChecked(False)
		self.checkBox15m.setChecked(False)
		self.checkBox17m.setChecked(False)
		self.checkBox20m.setChecked(False)
		self.checkBox30m.setChecked(False)
		self.checkBox40m.setChecked(False)
		self.checkBox80m.setChecked(False)
		self.checkBox160m.setChecked(False)
		self.checkBox5Band.setChecked(False)
		self.checkBox5B6mE.setChecked(False)
		self.checkBox5B12mE.setChecked(False)
		self.checkBox5B17mE.setChecked(False)
		self.checkBox5B30mE.setChecked(False)
		self.checkBox5B160mE.setChecked(False)
		self.checkBoxMixedHR.setChecked(False)
		self.checkBoxMixed.setChecked(False)
		self.checkBoxPhone.setChecked(False)
		self.checkBoxSatellite.setChecked(False)
		self.checkBoxDigitalHR.setChecked(False)
		self.checkBoxCW.setChecked(False)
		self.checkBoxCWHR.setChecked(False)
		self.checkBoxPhoneHR.setChecked(False)
		self.checkBoxChallenge.setChecked(False)
		self.checkBoxDigital.setChecked(False)

	def setAwards(self, ARRLAwards):
		self.clearAllAwards()
		Credits = ARRLAwards.getCredits()
		if "DXCC-2" in Credits:
			self.checkBox2m.setChecked(True)
		if "DXCC-6" in Credits:
			self.checkBox6m.setChecked(True)
		if "DXCC-10" in Credits:
			self.checkBox10m.setChecked(True)
		if "DXCC-12" in Credits:
			self.checkBox12m.setChecked(True)
		if "DXCC-15" in Credits:
			self.checkBox15m.setChecked(True)
		if "DXCC-17" in Credits:
			self.checkBox17m.setChecked(True)
		if "DXCC-20" in Credits:
			self.checkBox20m.setChecked(True)
		if "DXCC-30" in Credits:
			self.checkBox30m.setChecked(True)
		if "DXCC-40" in Credits:
			self.checkBox40m.setChecked(True)
		if "DXCC-80" in Credits:
			self.checkBox80m.setChecked(True)
		if "DXCC-160" in Credits:
			self.checkBox160m.setChecked(True)
		if "DXCC-5B" in Credits:
			self.checkBox5Band.setChecked(True)
		if "DXCC-5B-6" in Credits:
			self.checkBox5B6mE.setChecked(True)
		if "DXCC-5B-12" in Credits:
			self.checkBox5B12mE.setChecked(True)
		if "DXCC-5B-17" in Credits:
			self.checkBox5B17mE.setChecked(True)
		if "DXCC-5B-30" in Credits:
			self.checkBox5B30mE.setChecked(True)
		if "DXCC-5B-160" in Credits:
			self.checkBox5B160mE.setChecked(True)
		if "DXCC-M" in Credits:
			self.checkBoxMixed.setChecked(True)
		if "DXCC-M-HR" in Credits:
			self.checkBoxMixedHR.setChecked(True)
		if "DXCC-PH" in Credits:
			self.checkBoxPhone.setChecked(True)
		if "DXCC-PH-HR" in Credits:
			self.checkBoxPhoneHR.setChecked(True)
		if "DXCC-CW" in Credits:
			self.checkBoxCW.setChecked(True)
		if "DXCC-CW-HR" in Credits:
			self.checkBoxCWHR.setChecked(True)
		if "DXCC-RTTY" in Credits:
			self.checkBoxDigital.setChecked(True)
		if "DXCC-RTTY-HR" in Credits:
			self.checkBoxDigitalHR.setChecked(True)
		if "DXCC-SAT" in Credits:
			self.checkBoxSatellite.setChecked(True)
		if "DXCC-CHAL" in Credits:
			self.checkBoxChallenge.setChecked(True)

		self.showAwardsSection()
	
	def showUIElements(self):
		self.labelNumberVerifiedText.show()
		self.labelNumberVerifiedCount.show()
		self.labelPrefix.show()
		self.labelVerified.show()
		self.lineEditPrefix.show()
		self.labelDXCCPrefixEnter.show()


	def hideAwardsSection(self):
		self.labelVerifiedAwards.hide()
		self.labelBandAwards.hide()
		self.checkBox2m.hide()
		self.checkBox6m.hide()
		self.checkBox10m.hide()
		self.checkBox12m.hide()
		self.checkBox15m.hide()
		self.checkBox17m.hide()
		self.checkBox20m.hide()
		self.checkBox30m.hide()
		self.checkBox40m.hide()
		self.checkBox80m.hide()
		self.checkBox160m.hide()
		self.label5Band.hide()
		self.checkBox5Band.hide()
		self.checkBox5B6mE.hide()
		self.checkBox5B12mE.hide()
		self.checkBox5B17mE.hide()
		self.checkBox5B30mE.hide()
		self.checkBox5B160mE.hide()
		self.labelModeAwards.hide()
		self.checkBoxMixedHR.hide()
		self.checkBoxMixed.hide()
		self.checkBoxPhone.hide()
		self.checkBoxSatellite.hide()
		self.checkBoxDigitalHR.hide()
		self.checkBoxCW.hide()
		self.checkBoxCWHR.hide()
		self.checkBoxPhoneHR.hide()
		self.checkBoxChallenge.hide()
		self.checkBoxDigital.hide()

	def showAwardsSection(self):
		self.labelVerifiedAwards.show()
		self.labelBandAwards.show()
		self.checkBox2m.show()
		self.checkBox6m.show()
		self.checkBox10m.show()
		self.checkBox12m.show()
		self.checkBox15m.show()
		self.checkBox17m.show()
		self.checkBox20m.show()
		self.checkBox30m.show()
		self.checkBox40m.show()
		self.checkBox80m.show()
		self.checkBox160m.show()
		self.label5Band.show()
		self.checkBox5Band.show()
		self.checkBox5B6mE.show()
		self.checkBox5B12mE.show()
		self.checkBox5B17mE.show()
		self.checkBox5B30mE.show()
		self.checkBox5B160mE.show()
		self.labelModeAwards.show()
		self.checkBoxMixedHR.show()
		self.checkBoxMixed.show()
		self.checkBoxPhone.show()
		self.checkBoxSatellite.show()
		self.checkBoxDigitalHR.show()
		self.checkBoxCW.show()
		self.checkBoxCWHR.show()
		self.checkBoxPhoneHR.show()
		self.checkBoxChallenge.show()
		self.checkBoxDigital.show()


class LoTWVerified:
	def __init__(self):
		self.Country = "No Country"
		self.Mode = "No Mode"
		self.Band = "No Band"
		self.DXCCId = "No Id"
		self.Deleted = False
		self.Credits = {}

	def setCountry(self, Cnty):
		self.Country = Cnty

	def getCountry(self):
		return self.Country

	def setMode(self, Mod):
		self.Mode = Mod

	def getMode(self):
		return self.Mode

	def setBand(self, Bnd):
		self.Band = Bnd

	def getBand(self):
		return self.Band

	def setDXCCId(self, dxcc):
		self.DXCCId = dxcc

	def getDXCCId(self):
		return self.DXCCId

	def setDeleted(self, delDXCC):
		self.Deleted = delDXCC

	def getDeleted(self):
		return self.Deleted

	# This is a dictionary, I get a string from the adi file
	def setCredits(self, ARRLCred):
		# need to break up the string, check if the award exists as a key
		# if not add the key.. otherwise move on to the next award in the string!
		credits_split = ARRLCred.split(',')
		for s in credits_split:
			if s not in self.Credits:
				self.Credits[s] = s				

	def getCredits(self):
		return self.Credits

	def mergeCredits(self, newCredits):
		self.Credits = {**self.Credits, **newCredits}



def parseArrlData(filename):
	eohTag = "<eoh>"
	eorTag = "<eor>"

	newRecordFlag = True
	CountryTag = "<COUNTRY:"
	ModeTag = "<APP_LoTW_MODEGROUP:"
	BandTag = "<BAND:"
	DXCCTag = "<DXCC:"
	DeletedTag = "<APP_LOTW_DELETED_ENTITY:"
	CreditTag = "<APP_LoTW_CREDIT_GRANTED:"

	Start = False;

	DXCCKey = ""
	Count = 0;

	try:
		f = open("DXCC_QSLs_20171109_184150.adi", encoding = 'utf-8')
		for line in f:
		# look for the <eoh> signals the start of records!
			if not Start and line[0:len(eohTag)] == eohTag[0:len(eohTag)]:
				Start = True;	# Don't really need this unless I wnat to 
				continue
#				elif not Start:		# Here if we want to process the header.			
#					print(line)		# might want to get the date in the file!

# When we get here we have read the first line past the <eoh> tag..  so we are starting a new record.
# parse out the information I need for this DXCC verification and store it in the dictionary.  If the 
# ADIF Number already exists then update the record with the new verification, could be a new mode, band
# i.e. it is a verification of a new award type granted for this record.  Accumulate all the awards into
# the ADIFKey used for the dictionary entry.

# The DXCCKey will be the ADIF for this record.  The other data will be stored in another dict?
	
			if newRecordFlag:
				NewRecord = LoTWVerified()
				newRecordFlag = False

			# Not really needed
			if line[0:len(CountryTag)] == CountryTag[0:len(CountryTag)]:
				index = line.index('>')
				# Strip off the return and make sure it is Uppercase!
				NewRecord.setCountry(line[index+1: ].rstrip().upper())	
	
			# Not really needed
			if line[0:len(ModeTag)] == ModeTag[0:len(ModeTag)]:
				index = line.index('>')
				# Strip off the return and make sure it is Uppercase!
				NewRecord.setMode(line[index+1: ].rstrip().upper())		

			# Not really needed
			if line[0:len(BandTag)] == BandTag[0:len(BandTag)]:
				index = line.index('>')
				# Strip off the return and make sure it is Uppercase!
				NewRecord.setBand(line[index+1: ].rstrip().upper())		

			if line[0:len(DXCCTag)] == DXCCTag[0:len(DXCCTag)]:
				index = line.index('>')
				# Strip off the return and make sure it is Uppercase!
				NewRecord.setDXCCId(line[index+1: ].rstrip().upper())		

			if line[0:len(CreditTag)] == CreditTag[0:len(CreditTag)]:
				index = line.index('>')
				# Strip off the return and make sure it is Uppercase!
				NewRecord.setCredits(line[index+1: ].rstrip().upper())		

			if line[0:len(DeletedTag)] == DeletedTag[0:len(DeletedTag)]:
				index = line.index('>')
				if line[index+1: ].rstrip().upper() == 'NO':
					NewRecord.setDeleted(False)		
				else:
					NewRecord.setDeleted(True)		

			# Have we reached the end of this record... if so set the flag to create
			# a new Record and stor or update teh current record we decoded. 				
			if line[0:len(eorTag)] == eorTag[0:len(eorTag)]:
				newRecordFlag = True
				if NewRecord.getDXCCId() not in DXCCDict:
					DXCCDict[NewRecord.getDXCCId()] = NewRecord		
					Count += 1;
				else:
					Credits = DXCCDict[NewRecord.getDXCCId()]
					# I used this to see if the awards were duplicated... they are! so... read below!
					#print("Stored Credits", Credits.getCredits())				
					#print("New Credits", NewRecord.getCredits())
					#key = input("PAK...")
					# Need to put only new credits into stored record.  
					# Have to see if a sting is already in the awards 
					# and if it is don't put it in again!  Only add a 
					# string if it is not already in the awards string!

					# I use a dictionary here. class meber function breaks 
					# up the strings in the new awards string and sees if 
					# the key is in the dictionary. If it is don't add it 
					# again, duh!  Otherwise add a new awards as a new key...  
					# Could be overkill but it works and is simple.

					#print("Current..", Credits.getCredits(), "\n") 
					Credits.mergeCredits(NewRecord.getCredits())
					#print("New..", NewRecord.getCredits(), "\n")								
					#print("Merged..", DXCCDict[NewRecord.getDXCCId()].getCredits(), "\n")
					#key = input("PAK...")

		#	for K in sorted(DXCCDict):					# Need to add the bands and modes worked...
		#		print (K)
	finally:
		f.close()
		return Count


if __name__ == "__main__":
	app = QtGui.QApplication(sys.argv)

	# Create and display the splash screen
#	splash_pix = QtGui.QPixmap('splash_loading.png')
#	splash = QtGui.QSplashScreen(splash_pix, QtCore.Qt.WindowStaysOnTopHint)
#	splash.setMask(splash_pix.mask())
#	splash.show()
#	app.processEvents()

	# Read in the DXCC list...  there is a problem with this and the ARRL verification
	# data.  The strings don't match 100%.  i.e. ARRL says UNITED STATES while DXCC list
	# says UNITED STATES OF AMERICA!  I have to use the ADIF number i.e. for US 291 to
	# do the matching, I think!  
	pl = plistlib.readPlist('cty_plist/cty.plist')

	# Parse the ARRL_DXCC_VERIFIED.ADI and creat our dictionary.
	ARRLcnt = parseArrlData("DXCC_QSLs_20171109_184150.adi")

	window = MyApp()
	window.show()
	window.setFixedSize(window.size())
	# blank out the Country Label
	window.labelNumberVerifiedCount.setText(str(ARRLcnt))
	window.labelCountry.setText("")
	window.labelPrefix.setText("")
	window.labelVerified.setText("")	

#	splash.finish(window)

	sys.exit(app.exec_())
